src/main/scala/com/microsoft/azure/cosmosdb/spark/rdd/CosmosDBRDDIterator.scala

```diff
@@ -360,62 +354,62 @@ class CosmosDBRDDIterator(hadoopConfig: mutable.Map[String, String],
+                  throw ex
+                }
+              case _ => ex.getCause match {
+                case docEx: DocumentClientException =>
+                  handleGoneException(connection, docEx)
+                  isGone = true
+                case _ =>
+                  logError(s"An IllegalStateException was thrown: ${ex.getMessage}")
               }
-            } else if (ex.getCause.isInstanceOf[DocumentClientException]) {
-              val docex = ex.getCause.asInstanceOf[DocumentClientException]
-              handleGoneException(connection, docex);
-              isGone = true
-            } else {
-              logError(s"An IllegalStateException was thrown: ${ex.getMessage}")
             }
-          case ex: Throwable => {
+          case ex: Throwable =>
             logError(s"UNHANDLED EXCEPTION: ${ex.getMessage}")
             throw ex
-          }
         }
       }
       iteratorDocument
     }
 
-    taskContext.addTaskFailureListener((context: TaskContext, ex: Throwable) => {
+    taskContext.addTaskFailureListener((_: TaskContext, ex: Throwable) => {
       logError("Handling Task Failure")
-      if (ex.isInstanceOf[DocumentClientException]) {
-        val dcx: DocumentClientException = ex.asInstanceOf[DocumentClientException]
-        if (dcx.getStatusCode == StatusCodes.SERVICE_UNAVAILABLE) {
-          logError("Service Unavailable")
-          connection.reinitializeClient(shouldReinitializeThroughput = false)
-        }
-      } else if (ex.isInstanceOf[IllegalStateException]
-        && ex.getCause != null
-        && ex.getCause.isInstanceOf[DocumentClientException]){
-        logError("Illegal State Exception with Service Unavailable")
-        val dcx: DocumentClientException = ex.getCause.asInstanceOf[DocumentClientException]
-        if (dcx.getStatusCode == StatusCodes.SERVICE_UNAVAILABLE) {
-          connection.reinitializeClient(shouldReinitializeThroughput = false)
-        }
+      ex match {
+        case dcx: DocumentClientException =>
+          if (dcx.getStatusCode == StatusCodes.SERVICE_UNAVAILABLE) {
+            logError("Service Unavailable")
+            connection.reinitializeClient()
+          }
+        case _: IllegalStateException if ex.getCause.isInstanceOf[DocumentClientException] && ex.getCause != null =>
+          logError("Illegal State Exception with Service Unavailable")
+          val dcx: DocumentClientException = ex.getCause.asInstanceOf[DocumentClientException]
+          if (dcx.getStatusCode == StatusCodes.SERVICE_UNAVAILABLE) {
+            connection.reinitializeClient()
+          }
+        case _ =>
       }
     })
 
     // Register an on-task-completion callback to close the input stream.
-    taskContext.addTaskCompletionListener((context: TaskContext) => {
-      connection.reinitializeClient(shouldReinitializeThroughput = false)
+    taskContext.addTaskCompletionListener((_: TaskContext) => {
+      connection.reinitializeClient()
```

Comment:
Is it necessarily to reinitialize the client after every ready?  This might actually be one of the big contributors to offer/db reads.

Reply:
The reinitializeClient won't trigger the metadata reads on offer/db - but it will trigger refreshing the partition key range cache. But you are right - it shouldn't be necessary here - it is done form computePartitions which is the right time for queries,